python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll

```diff
@@ -0,0 +1,1540 @@
+}
+
+/**
+ * Gets a potential definition of the function `name` according to our approximation of
+ * MRO for the class `cls` (see `getNextClassInMro` for more information).
+ */
+Function findFunctionAccordingToMro(Class cls, string name) {
+  result = cls.getAMethod() and
+  result.getName() = name
+  or
+  not exists(Function f | f.getName() = name and f = cls.getAMethod()) and
+  result = findFunctionAccordingToMro(getNextClassInMro(cls), name)
+}
+
+/**
+ * Gets a class that, from an approximated MRO calculation, might be the next class
+ * after `cls` in the MRO for `startingClass`.
+ *
+ * Note: this is almost the same as `getNextClassInMro`, except we know the
+ * `startingClass`, which can give slightly more precise results.
+ *
+ * See QLDoc for `getNextClassInMro`.
+ */
+Class getNextClassInMroKnownStartingClass(Class cls, Class startingClass) {
+  cls.getBase(0) = classTracker(result).asExpr() and
+  cls = getADirectSuperclass*(startingClass)
+  or
+  exists(Class sub, int i | sub = getADirectSuperclass*(startingClass) |
+    sub.getBase(i) = classTracker(cls).asExpr() and
+    sub.getBase(i + 1) = classTracker(result).asExpr() and
+    not result = cls
+  )
+}
+
+private Function findFunctionAccordingToMroKnownStartingClass(
+  Class cls, Class startingClass, string name
+) {
+  result = cls.getAMethod() and
+  result.getName() = name and
+  cls = getADirectSuperclass*(startingClass)
+  or
+  not exists(Function f | f.getName() = name and f = cls.getAMethod()) and
+  result =
+    findFunctionAccordingToMroKnownStartingClass(getNextClassInMroKnownStartingClass(cls,
+        startingClass), startingClass, name)
+}
+
+/**
+ * Gets a potential definition of the function `name` according to our approximation of
+ * MRO for the class `cls` (see `getNextClassInMroKnownStartingClass` for more information).
+ *
+ * Note: this is almost the same as `findFunctionAccordingToMro`, except we know the
+ * `startingClass`, which can give slightly more precise results.
+ */
+pragma[inline]
+Function findFunctionAccordingToMroKnownStartingClass(Class startingClass, string name) {
+  result = findFunctionAccordingToMroKnownStartingClass(startingClass, startingClass, name)
+}
+
+// =============================================================================
+// attribute trackers
+// =============================================================================
+/** Gets a reference to the attribute read `attr` */
+private TypeTrackingNode attrReadTracker(TypeTracker t, AttrRead attr) {
+  t.start() and
+  result = attr and
+  attr.getObject() in [
+      classTracker(_), classInstanceTracker(_), selfTracker(_), clsTracker(_),
+      superCallNoArgumentTracker(_), superCallTwoArgumentTracker(_, _)
+    ]
+  or
+  exists(TypeTracker t2 | result = attrReadTracker(t2, attr).track(t2, t))
+}
+
+/** Gets a reference to the attribute read `attr` */
+Node attrReadTracker(AttrRead attr) { attrReadTracker(TypeTracker::end(), attr).flowsTo(result) }
```

Comment:
I don't think so no, because of the the restriction here

```
  attr.getObject() in [
      classTracker(_), classInstanceTracker(_), selfTracker(_), clsTracker(_),
      superCallNoArgumentTracker(_), superCallTwoArgumentTracker(_, _)
    ]
```

(I actually had one `AttrRead` tracker for each of these initially, but grouped them together at some point)

Reply:
Right, to keep it here is correct.